components/butterfly/src/rumor/service_config.rs

```diff
@@ -102,16 +104,23 @@ impl ServiceConfig {
         Ok(())
     }
 
-    pub fn config(&self) -> Result<String> {
-        if self.get_encrypted() {
+    pub fn config(&self) -> Result<toml::Value> {
+        let config = if self.get_encrypted() {
             let bytes = try!(BoxKeyPair::decrypt(self.get_config(), &default_cache_key_path(None)));
-            let config = try!(String::from_utf8(bytes).map_err(Error::ServiceConfigNotUtf8));
-            Ok(config)
+            let encoded = try!(str::from_utf8(&bytes)
+                .map_err(|e| Error::ServiceConfigNotUtf8(self.get_service_group().to_string(), e)));
+            try!(self.parse_config(&encoded))
         } else {
-            let config = try!(String::from_utf8(self.get_config().to_vec())
-                .map_err(Error::ServiceConfigNotUtf8));
-            Ok(config)
-        }
+            let encoded = try!(str::from_utf8(self.get_config())
+                .map_err(|e| Error::ServiceConfigNotUtf8(self.get_service_group().to_string(), e)));
+            try!(self.parse_config(&encoded))
+        };
+        Ok(config)
+    }
+
+    fn parse_config(&self, encoded: &str) -> Result<toml::Value> {
+        toml::Value::from_str(encoded)
+            .map_err(|e| Error::ServiceConfigDecode(self.get_service_group().to_string(), e))
```

Comment:
Feel free to ignore manager meddling: This just hit the rule of three.  Would a helper function be useful here?